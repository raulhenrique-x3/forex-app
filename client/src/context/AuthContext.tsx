import React, { createContext, useMemo, useState } from "react";

interface IProps {
  isAuthenticated: boolean;
  login: (value: boolean) => void;
  logout: () => void;
}

export const AuthContext = createContext<IProps>({
  isAuthenticated: false,
  login() {},
  logout() {},
});

interface IChildren {
  children: React.ReactNode;
}

export function AuthContextProvider({ children }: IChildren) {
  const [isAuthenticated, setIsAuthenticated] = useState<boolean>(false);

  const login = (value: boolean) => {
    setIsAuthenticated(value);
  };

  const logout = () => {
    setIsAuthenticated(false);
  };

  const value = useMemo(() => ({ isAuthenticated, login, logout }), [isAuthenticated]);
  return <AuthContext.Provider value={value}>{children}</AuthContext.Provider>;
}

export const AuthConsumer = AuthContext.Consumer;
