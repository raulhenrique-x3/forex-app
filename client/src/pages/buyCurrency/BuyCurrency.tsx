import { useEffect, useState } from "react";
import { Line } from "react-chartjs-2";
import {
  Text,
  Select,
  Container,
  Input,
  Button,
  InputRightElement,
  CheckboxIcon,
  InputGroup,
  InputLeftElement,
  Flex,
  useToast,
} from "@chakra-ui/react";
import {
  Chart as ChartJS,
  CategoryScale,
  LinearScale,
  PointElement,
  LineElement,
  Title,
  Tooltip,
  Legend,
} from "chart.js";
import Header from "../../components/header/Header";
import { useParams } from "react-router";
import axios from "axios";
import { CurrencyWallet } from "../../components/currencyWallet/CurrencyWallet";
import { IUser } from "../../interface/interface";
import { Wallet } from "../../components/wallet/Wallet";
import styles from "./buyCurrency.module.scss";
import io from "socket.io-client";

ChartJS.register(CategoryScale, LinearScale, PointElement, LineElement, Title, Tooltip, Legend);

export const options = {
  responsive: true,
  plugins: {
    legend: {
      position: "top" as const,
    },
    title: {
      display: true,
    },
  },
};

interface IBuyCurrency {
  map(arg0: (i: any) => any): any;
  source?: string;
  target?: string;
  value?: number;
  time?: number;
}

export const BuyCurrency = () => {
  const socket = io("ws://localhost:5000", { autoConnect: true });
  const [apiData, setApiData] = useState<IBuyCurrency[]>([]);
  const [userData, setUserData] = useState<IUser[]>([]);
  const [currencyValue, setCurrencyValue] = useState<IBuyCurrency[]>([]);
  const [userQuantity, setUserQuantity] = useState<string>();
  const [choosedCurrency, setChoosedCurrency] = useState<string>();
  const [ioResponse, setIoResponse] = useState<object[]>();
  const toast = useToast();

  let { userId } = useParams();
  const { exchange } = useParams();

  useEffect(() => {
    socket.on("money_deposited", (data) => {
      setIoResponse([data]);
    });

    return () => {
      socket.off("money_deposited");
    };
  }, [socket]);

  useEffect(() => {
    axios
      .get(`http://localhost:5000/home/${userId}`)
      .then((res) => {
        setUserData([res.data]);
      })
      .catch((error) => {
        console.error(error);
      });
  }, [ioResponse]);

  useEffect(() => {
    axios
      .get(`http://localhost:5000/api/history/${exchange}`)
      .then((res) => {
        setApiData(res.data);
      })
      .catch((err) => console.error(err));
  }, []);

  useEffect(() => {
    axios
      .get(`http://localhost:5000/api/${exchange}`)
      .then((res) => {
        setCurrencyValue(res.data);
      })
      .catch((err) => console.error(err));
  }, []);

  let numbers: number[] = [];

  for (let i = 0; i < 169; i++) {
    numbers.push(i);
  }

  const data = {
    labels: numbers,
    datasets: [
      {
        label: "Taxa de CÃ¢mbio do GBP",
        borderColor: "rgb(255, 99, 132)",
        backgroundColor: "rgba(255, 99, 132, 0.5)",
        data: apiData[0]?.map((i) => i.value),
      },
    ],
  };

  const buyCurrency = () => {
    socket.emit("deposit_money", { depositMoney: parseFloat(userQuantity!) });
    axios
      .put(`http://localhost:5000/exchange/${userId}`, {
        currencyValue: currencyValue[0]?.value,
        userQuantity: parseFloat(userQuantity!),
        choosedCurrency: choosedCurrency,
      })
      .then((res) => {
        axios.put(`http://localhost:5000/history/${userId}`, {
          purchasedCurrency: choosedCurrency,
          currencyAmount: parseFloat(userQuantity!),
        });
        toast({
          title: res.data,
          status: "success",
          isClosable: true,
        });
      })
      .catch((err) => {
        toast({
          title: err.message,
          status: "error",
          isClosable: true,
        });
      });
  };

  return (
    <main className={styles.buyCurrencyMain}>
      <Header user={userData} />
      <div className={styles.desktopCurrencyWallet}>
        <Line options={options} data={data} />

        <Container display={"flex"} flexDirection={"column"} gap={4}>
          <Container>
            <Wallet ioUpdate={ioResponse} />
            <CurrencyWallet walletName="GBP Wallet" user={userData} />
            <CurrencyWallet walletName="USD Wallet" user={userData} />
          </Container>
          <Container display={"flex"} flexDirection={"column"} gap={4}>
            <Text fontSize="xl">Buy</Text>
            <InputGroup>
              <InputLeftElement pointerEvents="none" color="gray.300" fontSize="1.2em" children="$" />
              <Input placeholder="Enter amount" onChange={(e) => setUserQuantity(e.target.value)} />
              <InputRightElement children={<CheckboxIcon color="green.500" />} />
            </InputGroup>
            <Select onChange={(e) => setChoosedCurrency(e.target.value)} width="100%">
              <option value={""}>Choose an currency to buy</option>
              {exchange === "usd_to_gbp" ? <option value="GBP" label="GBP" /> : <option value="USD" label="USD" />}
            </Select>
            {userQuantity === "NaN" || userQuantity === "" || userQuantity === undefined ? (
              <></>
            ) : (
              <Text fontSize={"xl"} color={"#000000"} fontWeight={"bold"}>
                Total: ${currencyValue[0]?.value! * parseFloat(userQuantity)}
              </Text>
            )}
            <Button onClick={buyCurrency} width={"100%"} backgroundColor={"#8236FD"} color={"#ffffff"}>
              Buy
            </Button>
          </Container>
        </Container>
      </div>
    </main>
  );
};
