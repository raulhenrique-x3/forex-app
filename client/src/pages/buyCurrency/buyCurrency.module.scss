.buyCurrencyMain {
  width: 100%;
}

.graphic {
  width: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
}

.graphicCurrency {
  display: flex;
  justify-content: center;
  align-items: center;
  width: 420px;
  height: 420px;
}

@media (min-width: 768px) {
  .desktopCurrencyWallet {
    width: 100%;
    height: 90vh;
    display: flex;
    justify-content: center;
    align-items: center;
  }

  .graphicCurrency {
    width: 700px;
    height: 700px;
  }
}
