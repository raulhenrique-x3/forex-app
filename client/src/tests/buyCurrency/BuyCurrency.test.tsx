import { ChakraProvider } from "@chakra-ui/react";
import { cleanup, render, screen, act, fireEvent, waitFor } from "@testing-library/react";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import { create } from "react-test-renderer";
import { AuthContextProvider } from "../../context/AuthContext";
import { BuyCurrency } from "../../pages/buyCurrency/BuyCurrency";
import "setimmediate";
import axios from "../../__mocks__/axios";

global.ResizeObserver = jest.fn().mockImplementation(() => ({
  observe: jest.fn(),
  unobserve: jest.fn(),
  disconnect: jest.fn(),
}));
afterEach(cleanup);

describe("He should be able to buy currency", () => {
  it("should be in the Buy Currency screen", () => {
    create(
      <AuthContextProvider>
        <ChakraProvider>
          <MemoryRouter initialEntries={["/wallet/buy_currency/gbp_to_usd/63e5135202704aea76e5fc52"]}>
            <Routes>
              <Route path={`/wallet/buy_currency/:exchange/:userId`} element={<BuyCurrency />} />
            </Routes>
          </MemoryRouter>
        </ChakraProvider>
      </AuthContextProvider>
    );
    const buyButton = screen.findByTestId("BuyCurrency");
    expect(buyButton).toBeTruthy();
  });

  it("should be able to buy currency if he sets the values and click on buy button", async () => {
    const buy = axios.put.mockResolvedValueOnce({
      currencyValue: 0.8,
      userQuantity: 10,
      choosedCurrency: "GBP-USD",
    });
    act(() => {
      render(
        <AuthContextProvider>
          <ChakraProvider>
            <MemoryRouter initialEntries={["/wallet/buy_currency/gbp_to_usd/63e5135202704aea76e5fc52"]}>
              <Routes>
                <Route path={`/wallet/buy_currency/:exchange/:userId`} element={<BuyCurrency jestBuyTest={buy} />} />
              </Routes>
            </MemoryRouter>
          </ChakraProvider>
        </AuthContextProvider>
      );
    });
    const buyButton = screen.findByTestId("BuyCurrency");
    expect(buyButton).toBeTruthy();
    fireEvent.click(await buyButton);
    expect(buy).toBeCalled();
  });

  it("should be receive an error when try to buy currency with wrong values", async () => {
    const buy = axios.put.mockResolvedValueOnce({
      currencyValue: 0.8,
      userQuantity: -10,
      choosedCurrency: "GBP-ERROR",
    });
    act(() => {
      render(
        <AuthContextProvider>
          <ChakraProvider>
            <MemoryRouter initialEntries={["/wallet/buy_currency/gbp_to_usd/63e5135202704aea76e5fc52"]}>
              <Routes>
                <Route path={`/wallet/buy_currency/:exchange/:userId`} element={<BuyCurrency jestBuyTest={buy} />} />
              </Routes>
            </MemoryRouter>
          </ChakraProvider>
        </AuthContextProvider>
      );
    });
    const buyButton = screen.findByTestId("BuyCurrency");
    expect(buyButton).toBeTruthy();
    const buyEv = fireEvent.click(await buyButton);
    console.log("Error buy ev", buyEv);
  });

  it("should be able to see the transition in the history", async () => {
    const buy = axios.put.mockResolvedValueOnce({
      purchasedCurrency: "GBP-USD",
      currencyAmount: 10,
    });
    act(() => {
      render(
        <AuthContextProvider>
          <ChakraProvider>
            <MemoryRouter initialEntries={["/wallet/buy_currency/gbp_to_usd/63e5135202704aea76e5fc52"]}>
              <Routes>
                <Route path={`/wallet/buy_currency/:exchange/:userId`} element={<BuyCurrency jestBuyTest={buy} />} />
              </Routes>
            </MemoryRouter>
          </ChakraProvider>
        </AuthContextProvider>
      );
    });
    const buyButton = screen.findByTestId("BuyCurrency");
    expect(buyButton).toBeTruthy();
    const buyEv = fireEvent.click(await buyButton);
    console.log("History event", buyEv);
  });
});
