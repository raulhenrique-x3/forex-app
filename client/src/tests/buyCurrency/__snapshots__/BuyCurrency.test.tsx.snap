// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuyCurrency should render the BuyCurrency component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <main
        class="buyCurrencyMain"
      >
        <header
          class="header"
        >
          <a
            href="/home/undefined"
          >
            <p
              class="headerText"
            >
              Forex-app
            </p>
          </a>
          <div
            class="desktopView"
          >
            <button
              class="chakra-button css-1aemnt7"
              type="button"
            >
              <svg
                color="#8236FD"
                fill="none"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                style="color: rgb(130, 54, 253);"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M4 6h16M4 12h16M4 18h16"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
            </button>
          </div>
          <nav
            class="desktopNav"
          >
            <ul
              class="desktopUl"
            >
              <li>
                <a
                  href="/home/undefined"
                >
                  Home 
                </a>
              </li>
              <li>
                <a
                  href="/wallet/undefined"
                >
                  Wallet 
                </a>
              </li>
              <li>
                <a
                  href="/history/undefined"
                >
                  History 
                </a>
              </li>
            </ul>
            <button
              class="chakra-button css-4xx2wk"
              data-testid="Logout"
              type="button"
            >
              Logout
            </button>
          </nav>
        </header>
        <p
          class="chakra-text css-phodh6"
        >
           $:
        </p>
        <div
          class="desktopCurrencyWallet"
        >
          <div
            class="graphic"
          >
            <div
              class="graphicCurrency"
            >
              <canvas
                height="0"
                role="img"
                style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
                width="0"
              />
            </div>
          </div>
          <div
            class="chakra-container css-1kkt86i"
          >
            <div
              class="chakra-container css-0"
            >
              <div
                class="chakra-card css-cb6d64"
                data-testid="walletComponent"
              >
                <div
                  class="css-70qvj9"
                >
                  <div
                    class="chakra-container css-0"
                  >
                    <p
                      class="chakra-text css-16equ04"
                    >
                      Wallet balance:
                       
                    </p>
                    <p
                      class="chakra-text css-1j1vrbk"
                    >
                      $NaN
                    </p>
                  </div>
                </div>
              </div>
              <div
                class="chakra-card css-1vppx1d"
              >
                <div
                  class="css-70qvj9"
                >
                  <div
                    class="chakra-container css-0"
                  >
                    <p
                      class="chakra-text css-1pwmtxc"
                    >
                      GBP Wallet
                    </p>
                    <p
                      class="chakra-text css-d5e2sn"
                    >
                      Total:
                       
                      $NaN
                    </p>
                  </div>
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 1024 1024"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm146 658c0 4.4-3.6 8-8 8H376.2c-4.4 0-8-3.6-8-8v-38.5c0-3.7 2.5-6.9 6.1-7.8 44-10.9 72.8-49 72.8-94.2 0-14.7-2.5-29.4-5.9-44.2H374c-4.4 0-8-3.6-8-8v-30c0-4.4 3.6-8 8-8h53.7c-7.8-25.1-14.6-50.7-14.6-77.1 0-75.8 58.6-120.3 151.5-120.3 26.5 0 51.4 5.5 70.3 12.7 3.1 1.2 5.2 4.2 5.2 7.5v39.5a8 8 0 0 1-10.6 7.6c-17.9-6.4-39-10.5-60.4-10.5-53.3 0-87.3 26.6-87.3 70.2 0 24.7 6.2 47.9 13.4 70.5h112c4.4 0 8 3.6 8 8v30c0 4.4-3.6 8-8 8h-98.6c3.1 13.2 5.3 26.9 5.3 41 0 40.7-16.5 73.9-43.9 91.1v4.7h180c4.4 0 8 3.6 8 8V722z"
                    />
                  </svg>
                </div>
              </div>
              <div
                class="chakra-card css-1vppx1d"
              >
                <div
                  class="css-70qvj9"
                >
                  <div
                    class="chakra-container css-0"
                  >
                    <p
                      class="chakra-text css-1pwmtxc"
                    >
                      USD Wallet
                    </p>
                    <p
                      class="chakra-text css-d5e2sn"
                    >
                      Total:
                       
                      $NaN
                    </p>
                  </div>
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 512 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M411.387 303.256c-3.119-9.577-7.891-18.561-14.301-26.952-6.422-8.382-14.396-15.826-23.93-22.331-9.539-6.498-20.721-11.63-33.553-15.4-5.143-1.363-14.189-3.506-26.104-6.418-8.516-2.074-16.5-4.2-25.5-6.367V120.065c9 2.396 15.252 6.202 21.926 10.43C324.204 139.535 333.157 155 335.78 176h69.174c-.654-18-4.65-32.76-11.996-46.02-8.07-14.543-18.977-27.024-32.73-36.956-13.75-9.922-30.225-17.49-48.377-22.455C303.967 68.416 297 66.605 288 65.386V32h-64v33.167c-7 1.044-15.148 2.445-22.426 4.25-17.242 4.283-32.388 10.868-45.951 19.764-13.571 8.905-24.352 20.112-32.604 33.627-8.251 13.523-12.312 29.52-12.312 48 0 9.585 1.407 18.993 4.157 28.235 2.752 9.241 7.442 17.967 14.042 26.181 6.603 8.214 15.495 15.658 26.687 22.332 11.183 6.672 24.705 12.064 41.576 16.171 9.287 2.345 18.83 4.534 26.83 6.576v119.586c-11-2.919-21.889-7.399-30.678-13.479-9.17-6.327-16.066-13.953-21.198-23.884-4.779-9.229-7.073-20.526-7.407-32.526H96c.695 21 5.25 39.494 13.672 55.371 8.799 16.604 20.533 29.96 35.204 40.562 14.662 10.613 31.393 18.356 51.198 23.491 8.885 2.304 18.926 3.96 27.926 5.23V480h64v-34.54c10-1.069 18.957-2.69 28.527-4.879 18.701-4.273 35.645-11.036 50.316-20.276 14.662-9.24 26.621-21.128 35.611-35.681 8.98-14.541 13.545-32.085 13.545-52.619.001-9.578-1.501-19.164-4.612-28.749zM224 209.699c-12-3.743-23.912-9.088-32.051-16.048-8.621-7.355-12.673-17.534-12.673-30.545 0-9.241 2.414-16.94 7.004-23.102 4.58-6.161 9.912-11.038 16.88-14.631 6.18-3.189 13.84-5.565 20.84-7.138v91.464zm118.902 149.772c-2.939 6.673-7.699 12.576-14.303 17.711-6.602 5.133-15.744 9.328-26.377 12.577-4.5 1.378-8.223 2.444-14.223 3.236v-107.11c10 2.624 18.18 5.332 26.326 8.131 8.062 2.744 15.748 7.443 22.537 14.116 6.785 6.676 10.309 17.03 10.309 31.06.001 6.85-1.337 13.605-4.269 20.279z"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <div
              class="chakra-container css-1kkt86i"
            >
              <p
                class="chakra-text css-6dvxm2"
              >
                Buy
              </p>
              <div
                class="chakra-input__group css-bx0blc"
              >
                <div
                  class="chakra-input__left-element css-fzewfr"
                >
                  $
                </div>
                <input
                  class="chakra-input css-0"
                  placeholder="Enter amount"
                />
                <div
                  class="chakra-input__right-element css-dp9r5f"
                />
              </div>
              <div
                class="chakra-select__wrapper css-qptiug"
              >
                <select
                  class="chakra-select css-1ik61og"
                >
                  <option
                    value=""
                  >
                    Choose an currency to buy
                  </option>
                  <option
                    label="USD"
                    value="USD-GBP"
                  />
                </select>
                <div
                  class="chakra-select__icon-wrapper css-162mkon"
                >
                  <svg
                    aria-hidden="true"
                    class="chakra-select__icon"
                    focusable="false"
                    role="presentation"
                    style="width: 1em; height: 1em; color: currentColor;"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                      fill="currentColor"
                    />
                  </svg>
                </div>
              </div>
              <button
                class="chakra-button css-s1zm2w"
                type="button"
              >
                Buy
              </button>
            </div>
          </div>
        </div>
      </main>
    </div>
  </body>,
  "container": <div>
    <main
      class="buyCurrencyMain"
    >
      <header
        class="header"
      >
        <a
          href="/home/undefined"
        >
          <p
            class="headerText"
          >
            Forex-app
          </p>
        </a>
        <div
          class="desktopView"
        >
          <button
            class="chakra-button css-1aemnt7"
            type="button"
          >
            <svg
              color="#8236FD"
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              style="color: rgb(130, 54, 253);"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M4 6h16M4 12h16M4 18h16"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
          </button>
        </div>
        <nav
          class="desktopNav"
        >
          <ul
            class="desktopUl"
          >
            <li>
              <a
                href="/home/undefined"
              >
                Home 
              </a>
            </li>
            <li>
              <a
                href="/wallet/undefined"
              >
                Wallet 
              </a>
            </li>
            <li>
              <a
                href="/history/undefined"
              >
                History 
              </a>
            </li>
          </ul>
          <button
            class="chakra-button css-4xx2wk"
            data-testid="Logout"
            type="button"
          >
            Logout
          </button>
        </nav>
      </header>
      <p
        class="chakra-text css-phodh6"
      >
         $:
      </p>
      <div
        class="desktopCurrencyWallet"
      >
        <div
          class="graphic"
        >
          <div
            class="graphicCurrency"
          >
            <canvas
              height="0"
              role="img"
              style="display: block; box-sizing: border-box; height: 0px; width: 0px;"
              width="0"
            />
          </div>
        </div>
        <div
          class="chakra-container css-1kkt86i"
        >
          <div
            class="chakra-container css-0"
          >
            <div
              class="chakra-card css-cb6d64"
              data-testid="walletComponent"
            >
              <div
                class="css-70qvj9"
              >
                <div
                  class="chakra-container css-0"
                >
                  <p
                    class="chakra-text css-16equ04"
                  >
                    Wallet balance:
                     
                  </p>
                  <p
                    class="chakra-text css-1j1vrbk"
                  >
                    $NaN
                  </p>
                </div>
              </div>
            </div>
            <div
              class="chakra-card css-1vppx1d"
            >
              <div
                class="css-70qvj9"
              >
                <div
                  class="chakra-container css-0"
                >
                  <p
                    class="chakra-text css-1pwmtxc"
                  >
                    GBP Wallet
                  </p>
                  <p
                    class="chakra-text css-d5e2sn"
                  >
                    Total:
                     
                    $NaN
                  </p>
                </div>
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 1024 1024"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm146 658c0 4.4-3.6 8-8 8H376.2c-4.4 0-8-3.6-8-8v-38.5c0-3.7 2.5-6.9 6.1-7.8 44-10.9 72.8-49 72.8-94.2 0-14.7-2.5-29.4-5.9-44.2H374c-4.4 0-8-3.6-8-8v-30c0-4.4 3.6-8 8-8h53.7c-7.8-25.1-14.6-50.7-14.6-77.1 0-75.8 58.6-120.3 151.5-120.3 26.5 0 51.4 5.5 70.3 12.7 3.1 1.2 5.2 4.2 5.2 7.5v39.5a8 8 0 0 1-10.6 7.6c-17.9-6.4-39-10.5-60.4-10.5-53.3 0-87.3 26.6-87.3 70.2 0 24.7 6.2 47.9 13.4 70.5h112c4.4 0 8 3.6 8 8v30c0 4.4-3.6 8-8 8h-98.6c3.1 13.2 5.3 26.9 5.3 41 0 40.7-16.5 73.9-43.9 91.1v4.7h180c4.4 0 8 3.6 8 8V722z"
                  />
                </svg>
              </div>
            </div>
            <div
              class="chakra-card css-1vppx1d"
            >
              <div
                class="css-70qvj9"
              >
                <div
                  class="chakra-container css-0"
                >
                  <p
                    class="chakra-text css-1pwmtxc"
                  >
                    USD Wallet
                  </p>
                  <p
                    class="chakra-text css-d5e2sn"
                  >
                    Total:
                     
                    $NaN
                  </p>
                </div>
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 512 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M411.387 303.256c-3.119-9.577-7.891-18.561-14.301-26.952-6.422-8.382-14.396-15.826-23.93-22.331-9.539-6.498-20.721-11.63-33.553-15.4-5.143-1.363-14.189-3.506-26.104-6.418-8.516-2.074-16.5-4.2-25.5-6.367V120.065c9 2.396 15.252 6.202 21.926 10.43C324.204 139.535 333.157 155 335.78 176h69.174c-.654-18-4.65-32.76-11.996-46.02-8.07-14.543-18.977-27.024-32.73-36.956-13.75-9.922-30.225-17.49-48.377-22.455C303.967 68.416 297 66.605 288 65.386V32h-64v33.167c-7 1.044-15.148 2.445-22.426 4.25-17.242 4.283-32.388 10.868-45.951 19.764-13.571 8.905-24.352 20.112-32.604 33.627-8.251 13.523-12.312 29.52-12.312 48 0 9.585 1.407 18.993 4.157 28.235 2.752 9.241 7.442 17.967 14.042 26.181 6.603 8.214 15.495 15.658 26.687 22.332 11.183 6.672 24.705 12.064 41.576 16.171 9.287 2.345 18.83 4.534 26.83 6.576v119.586c-11-2.919-21.889-7.399-30.678-13.479-9.17-6.327-16.066-13.953-21.198-23.884-4.779-9.229-7.073-20.526-7.407-32.526H96c.695 21 5.25 39.494 13.672 55.371 8.799 16.604 20.533 29.96 35.204 40.562 14.662 10.613 31.393 18.356 51.198 23.491 8.885 2.304 18.926 3.96 27.926 5.23V480h64v-34.54c10-1.069 18.957-2.69 28.527-4.879 18.701-4.273 35.645-11.036 50.316-20.276 14.662-9.24 26.621-21.128 35.611-35.681 8.98-14.541 13.545-32.085 13.545-52.619.001-9.578-1.501-19.164-4.612-28.749zM224 209.699c-12-3.743-23.912-9.088-32.051-16.048-8.621-7.355-12.673-17.534-12.673-30.545 0-9.241 2.414-16.94 7.004-23.102 4.58-6.161 9.912-11.038 16.88-14.631 6.18-3.189 13.84-5.565 20.84-7.138v91.464zm118.902 149.772c-2.939 6.673-7.699 12.576-14.303 17.711-6.602 5.133-15.744 9.328-26.377 12.577-4.5 1.378-8.223 2.444-14.223 3.236v-107.11c10 2.624 18.18 5.332 26.326 8.131 8.062 2.744 15.748 7.443 22.537 14.116 6.785 6.676 10.309 17.03 10.309 31.06.001 6.85-1.337 13.605-4.269 20.279z"
                  />
                </svg>
              </div>
            </div>
          </div>
          <div
            class="chakra-container css-1kkt86i"
          >
            <p
              class="chakra-text css-6dvxm2"
            >
              Buy
            </p>
            <div
              class="chakra-input__group css-bx0blc"
            >
              <div
                class="chakra-input__left-element css-fzewfr"
              >
                $
              </div>
              <input
                class="chakra-input css-0"
                placeholder="Enter amount"
              />
              <div
                class="chakra-input__right-element css-dp9r5f"
              />
            </div>
            <div
              class="chakra-select__wrapper css-qptiug"
            >
              <select
                class="chakra-select css-1ik61og"
              >
                <option
                  value=""
                >
                  Choose an currency to buy
                </option>
                <option
                  label="USD"
                  value="USD-GBP"
                />
              </select>
              <div
                class="chakra-select__icon-wrapper css-162mkon"
              >
                <svg
                  aria-hidden="true"
                  class="chakra-select__icon"
                  focusable="false"
                  role="presentation"
                  style="width: 1em; height: 1em; color: currentColor;"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                    fill="currentColor"
                  />
                </svg>
              </div>
            </div>
            <button
              class="chakra-button css-s1zm2w"
              type="button"
            >
              Buy
            </button>
          </div>
        </div>
      </div>
    </main>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
